### PR TITLE
CP-20287: expose experimental features to the API

### PR DESCRIPTION
**Pull request by kc284**

### CODE REVIEW

**Comment by kc284** [comment]
Note: this PR should be merged at the same time with xapi-project/xen-api/PR#2932 and xs/v6/PR#7

File: v6/v6_interface.ml [hunk 1]
```diff
@@ -28,9 +28,9 @@ exception Missing_connection_details
 exception License_checkout_error of string
 
 (* dbg_str -> requested edition -> current params ->
-   new edition * new xapi params * new additional params *)
```
Comment (lindig) [inline v6/v6_interface.ml]: As mentioned in another review, this is fine for now but we should change this type to a record in the long run.  `string * (string * string) list * (string * string) list * (string * bool) list` is not a descriptive type and there is a good chance of putting strings into the wrong list.

Comment (robhoes) [inline v6/v6_interface.ml]: Well, let's do that right now then, otherwise it will not happen. I agree that using a record will make the interface clearer.

**Comment by robhoes** [comment]
> Removed files from version control as they are generated at compilation time.

Is that true? I don't see any `oasis setup` in the spec file.

**Comment by kc284** [comment]
I've reverted the commit that removed those files - my bad, I had done oasis setup in my build environment. Athough I would find it preferrable to move it to the spec file, I'm going to leave it out of this PR for now as it's out of scope.

File: v6/v6_interface.ml [hunk 2]
```diff
@@ -20,17 +20,22 @@ let uri () = "file:" ^ !default_path
 
 type debug_info = string
 
+type edition_info = {
```
Comment (lindig) [inline v6/v6_interface.ml]: The record introduces clarity - good. I would have called the type simply `edition` or `license` unless there is a type of that name already in use (outside this module). If this record provides details to an existing license, `license_attributes` or `license_attr` would be less generic than `info`. Anyway, I'm fine with keeping it.

Comment (robhoes) [inline v6/v6_interface.ml]: It does contain a couple of different things, but all related to the edition that was applied. I'm fine with `edition_info`.

File: v6/v6_interface.ml [hunk 3]
```diff
@@ -20,17 +20,22 @@ let uri () = "file:" ^ !default_path
...
+    edition: string;
+    licence_restrictions: (string * string) list;
```
Comment (robhoes) [inline v6/v6_interface.ml]: This actually contains more than just the restrictions. There are other parameters as well, that end up in `host.license_params`. Let's call it `xapi_params`, like in the old code comment. And one below can be `additional_params`.

Comment (robhoes) [inline v6/v6_interface.ml]: Actually, it looks like the one below is the one that contains more than just restriction. But the same argument applies.

Comment (robhoes) [inline v6/v6_interface.ml]: Actually, it looks like the split between these two lists is not really needed anymore. But we can look at simplifying that another time.

**Pull request merged**